### PR TITLE
Feature/import export catalogi

### PR DESCRIPTION
**Pull request by stevenbal**
fixes https://github.com/open-zaak/open-zaak/issues/189

exporting ResultaatType doesn't work yet, due to issues with the serializer: https://github.com/open-zaak/open-zaak/issues/256

### CODE REVIEW

File: src/openzaak/components/catalogi/tests/admin/test_import_export_catalogus.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        form = response.form
+        with open(self.filename, "rb") as f:
+            form["file"] = (
+                self.filename,
+                f.read(),
+            )
+            response = form.submit("_import")
+
+        self.assertIn(
+            "validation error occurred while deserializing a Catalogus", response.text
```
Comment (joeribekker) [inline src/openzaak/components/catalogi/tests/admin/test_import_export_catalogus.py]: This can fail if we translate this string.

Use the exact same translatable string as you did when raising the exception.

File: src/openzaak/templates/admin/change_list.html [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  {% comment %} {% if has_add_permission %}
+  {<li>
+    {% url cl.opts|admin_urlname:'add' as add_url %}
+    <a href="{% add_preserved_filters add_url is_popup to_field %}" class="addlink">
+      {% blocktrans with cl.opts.verbose_name as name %}Add {{ name }}{% endblocktrans %}
+    </a>
+  </li
+  {% endif %} {% endcomment %}
+  {{ block.super }}
+    {% if cl.opts.verbose_name == "catalogus" %}
```
Comment (joeribekker) [inline src/openzaak/templates/admin/change_list.html]: Since this code is only for the catalogus admin, I prefer making a new (copy) template file and point to it from the admin class (something like `template_name=...`) instead of using an if-statement.

File: src/openzaak/templates/admin/submit_line.html [hunk 3]
```diff
@@ -6,5 +6,9 @@
         <input type="submit" value="{% trans 'Publish' %}" name="_publish"/>
     {% endif %}
 
+    {% if original.domein %}
+        <input type="submit" value="{% trans 'Export' %}" name="_export"/>
```
Comment (joeribekker) [inline src/openzaak/templates/admin/submit_line.html]: This becomes messy now with all the actions and I definitly don't like the if-statement nor the attribute checking. But you can leave it for now and we'll see how to make it more elegant afterwards. I made #268 for this.

File: src/openzaak/templates/import_catalogus.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{% extends "admin/base_site.html" %}
```
Comment (joeribekker) [inline src/openzaak/templates/import_catalogus.html]: Can you move this file to template/admin/catalogi/?

File: src/openzaak/components/catalogi/management/commands/export.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+            setattr(request, "versioning_scheme", URLPathVersioning())
+            setattr(request, "version", "1")
+
+            for instance in objects:
+                serialized = serializer(instance, context={"request": request})
+                data = serialized.data
+
+                if resource == "BesluitType":
+                    data["zaaktypen"] = []
```
Comment (joeribekker) [inline src/openzaak/components/catalogi/management/commands/export.py]: I remember we spoke of this, but why is this here again?

I think a code-comment should explain why this is here.

File: src/openzaak/components/catalogi/admin/catalogus.py [hunk 6]
```diff
@@ -52,6 +64,94 @@ class CatalogusAdmin(ListObjectActionsAdminMixin, admin.ModelAdmin):
...
+                id_list.append([",".join(str(id) for id in ids)])
+
+        return resource_list, id_list
+
+    def response_post_save_change(self, request, obj):
+        if "_export" in request.POST:
+            # Clear messages
+            storage = messages.get_messages(request)
+            for i in storage:
+                pass
```
Comment (joeribekker) [inline src/openzaak/components/catalogi/admin/catalogus.py]: Can you check how @annashamray did this for creating a new version?

I can't remember this was needed there, but then again, it might just show 2 messages...heh.

File: src/openzaak/components/catalogi/management/commands/export.py (hunk 5 above)
Comment (stevenbal) [reply src/openzaak/components/catalogi/management/commands/export.py]: This is needed, because when exporting, `BesluitType.zaaktypen` and `ZaakType.besluittypen` both contain references, but then creating the `BesluitType` first will fail, because the `ZaakType` does not exist yet. I'll add a comment for it

File: src/openzaak/components/catalogi/api/serializers/relatieklassen.py [hunk 7]
```diff
@@ -9,7 +9,7 @@
 from ..validators import ZaakInformatieObjectTypeCatalogusValidator
 
 
-class ZaakTypeInformatieObjectTypeSerializer(serializers.HyperlinkedModelSerializer):
+class ZaakInformatieobjectTypeSerializer(serializers.HyperlinkedModelSerializer):
```
[the remark this answers is not in the document]
Comment (stevenbal) [reply src/openzaak/components/catalogi/api/serializers/relatieklassen.py]: Yes, but the model is named `ZaakInformatieobjectType` and I use the model name to retrieve the serializer. I suppose I could also rename the model from `ZaakInformatieobjectType` -> `ZaakTypeInformatieObjectType`.

File: src/openzaak/components/catalogi/admin/catalogus.py (hunk 6 above)
Comment (stevenbal) [reply src/openzaak/components/catalogi/admin/catalogus.py]: Yes, if I remove it, it will show 2 messages: the custom one I added and the original message that says that the resource was successfully edited or whatever

File: src/openzaak/components/catalogi/api/serializers/relatieklassen.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (joeribekker) [reply src/openzaak/components/catalogi/api/serializers/relatieklassen.py]: Please do.

**Comment by joeribekker** [comment]
Thanks, please do the rename as suggested here: https://github.com/open-zaak/open-zaak/pull/259#discussion_r350858787

File: src/openzaak/components/catalogi/admin/mixins.py [hunk 8]
```diff
@@ -122,3 +133,101 @@ def response_change(self, request, obj):
...
+                    "pk", flat=True
+                )
+            )
+
+        resource_list = []
+        id_list = []
+        for resource, ids in resources.items():
+            if ids:
+                resource_list.append([resource])
+                id_list.append([",".join(str(id) for id in ids)])
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/admin/mixins.py]: can't you pass an actual python list instead of turning it into a comma-separated list? This is passed to the management command args right?

File: src/openzaak/components/catalogi/management/commands/export.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            action="append",
+            nargs="*",
+            help=_("Name of the resource to export objects for"),
+        )
+        parser.add_argument(
+            "--ids",
+            help=_("IDs of objects to be exported for the resource"),
+            action="append",
+            nargs="*",
+            type=str,
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/management/commands/export.py]: I'd expect `type=int` because they are PKs?

File: src/openzaak/components/catalogi/management/commands/export.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    def handle(self, *args, **options):
+        archive_name = options.pop("archive_name")
+        resources = options.pop("resource")
+        ids = options.pop("ids")
+
+        filenames = []
+
+        for resource, id_list in zip(resources, ids):
+            resource = resource[0]
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/management/commands/export.py]: that's kinda weird?

File: src/openzaak/components/catalogi/management/commands/export.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+            serializer = getattr(serializers, f"{resource}Serializer")
+            objects = model.objects.filter(id__in=parsed_ids)
+
+            factory = APIRequestFactory()
+            request = factory.get("/")
+
+            setattr(request, "versioning_scheme", URLPathVersioning())
+            setattr(request, "version", "1")
+
+            for instance in objects:
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/management/commands/export.py]: can't this be replaced by a `serializer(instance=objects, many=True)` call?

File: src/openzaak/components/catalogi/management/commands/import.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    "ZaakInformatieobjectType",
+    "ResultaatType",
+    "RolType",
+    "StatusType",
+    "Eigenschap",
+]
+
+
+class Command(BaseCommand):
+    help = _("Import Catalogi data from a .zip file")
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/management/commands/import.py]:
```suggestion
    help = "Import Catalogi data from a .zip file"
```

File: src/openzaak/components/catalogi/management/commands/import.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+                    data = zip_file.read(f"{resource}.json").decode()
+                    for old, new in uuid_mapping.items():
+                        data = data.replace(old, new)
+
+                    data = json.loads(data)
+
+                    model = apps.get_model("catalogi", resource)
+                    serializer = getattr(serializers, f"{resource}Serializer")
+
+                    factory = APIRequestFactory()
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/management/commands/import.py]: stuff like this can be outside of the for-loop, right?

File: src/openzaak/components/catalogi/tests/admin/test_import_export_catalogus.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        )
+        response = form.submit("_import")
+
+        self.assertIn(
+            _("A validation error occurred while deserializing a Catalogus"),
+            response.text,
+        )
+        self.assertEqual(Catalogus.objects.count(), 1)
+
+    def tearDown(self):
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/tests/admin/test_import_export_catalogus.py]: please put these methods at the top of the class

also - it's better to not define `def tearDown`, but add it to the cleanup in `def setUp`:

```python
self.addCleanup(lambda: os.remove(self.filename))
```

Cleanup always runs, even if errors happen. tearDown doesn't always run

File: src/openzaak/components/catalogi/tests/management/test_import_export.py [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+
+class ExportCatalogiTests(TestCase):
+    def setUp(self):
+        self.filepath = os.path.join(PATH, "export_test.zip")
+
+    def test_export_catalogus(self):
+        catalogus = CatalogusFactory.create()
+
+        call_command(
+            "export", self.filepath, resource=[["Catalogus"]], ids=[[str(catalogus.id)]]
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/tests/management/test_import_export.py]: the arguments are weird, should be fixed in the command itself I think and then adapt the calling code

```suggestion
            "export", self.filepath, resource=["Catalogus"], ids=[[catalogus.id]]
```

Command should probably also validate that there are equal amounts of elements in `resource` as in `ids`

File: src/openzaak/templates/admin/submit_line.html (hunk 3 above)
Comment (joeribekker) [reply src/openzaak/templates/admin/submit_line.html]: Since #268 was merged, this branch will have merge conflicts with this change.

You can copy (in master branch) `templates/admin/catalogi/change_form_zaaktype.html` to `change_form_catalog.html` and use the same template logic to add this button without the if-statement.

Then use this template by specifying `change_form_template = "admin/catalogi/change_form_catalog.html"` in the 
`CatalogusAdmin` class.

**Comment by stevenbal** [comment]
@sergei-maertens I have changed it a bit, now when importing a zip, the file is not saved and the content is passed to the import management command. Similarly for exporting, I now pass the response to the export command and write the output to that response, instead of writing it to a file, then sending that file and then deleting that file

File: src/openzaak/components/catalogi/management/commands/export.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+            nargs="*",
+            type=int,
+        )
+
+    def handle(self, *args, **options):
+        archive_name = options.pop("archive_name")
+        response = options.pop("response")
+
+        if response and archive_name:
+            raise CommandError(
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/management/commands/export.py]: for a future exercise, but just informing you - with argparse you can define mutually exclusive options. https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_mutually_exclusive_group

File: src/openzaak/components/catalogi/openapi.yaml [hunk 17]
```diff
@@ -5887,7 +5887,7 @@ paths:
         format: uuid
   /zaaktype-informatieobjecttypen:
     get:
-      operationId: zaakinformatieobjecttype_list
+      operationId: zaaktypeinformatieobjecttype_list
```
Comment (sergei-maertens) [inline src/openzaak/components/catalogi/openapi.yaml]: that's a breaking change - clients operating on `operationId` (of which gemma-zds-clients is one) will crash on this change.

The schema introspection has a `get_operation_id` method, probably we should override it to keep the old one here.
